File: clients/nodejs/managed_smoke.js
"use strict";

// Run with: KUTTIDB_SERVER=/path/to/kuttidb node managed_smoke.js DATA_DIR [tcp]
const { Client } = require("./kuttidb_client");
const net = require("net");

function freeLoopbackPort() {
  return new Promise((resolve, reject) => {
    const server = net.createServer();
    server.once("error", reject);
    server.listen(0, "127.0.0.1", () => {
      const { port } = server.address();
      server.close((error) => error ? reject(error) : resolve(port));
    });
  });
}

async function main() {
  const dataDir = process.argv[2];
  if (!dataDir) throw new Error("managed data directory required");
  const tcp = process.argv[3] === "tcp";
  const port = tcp ? await freeLoopbackPort() : undefined;
  const client = await Client.managed({
    dataDir,
    executable: process.env.KUTTIDB_SERVER,
    idleTimeout: 0.25,
    startupTimeout: 5,
    transport: tcp ? "tcp" : "unix",
    host: tcp ? "127.0.0.1" : undefined,
    port,
  });
  try {
    await client.put("managed-node", Buffer.from("value"));
    const value = await client.get("managed-node");
    if (!value || value.toString() !== "value") throw new Error("managed Node value mismatch");
  } finally {
    await client.close();
  }
}

main().catch((error) => { console.error(error); process.exit(1); });

File: clients/java/ManagedSmoke.java
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.net.ServerSocket;

/** Minimal opt-in managed lifecycle smoke test. */
public final class ManagedSmoke {
    public static void main(String[] args) throws Exception {
        if (args.length != 2 && args.length != 3) throw new IllegalArgumentException("data directory, executable, and optional tcp transport required");
        boolean tcp = args.length == 3 && "tcp".equals(args[2]);
        int port = 7379;
        if (tcp) try (ServerSocket reservation = new ServerSocket(0)) { port = reservation.getLocalPort(); }
        KuttiDBClient.ManagedServerOptions options = new KuttiDBClient.ManagedServerOptions(
                Path.of(args[0]), Path.of(args[1]), 250, 5000, null,
                tcp ? "tcp" : "unix", "127.0.0.1", port);
        try (KuttiDBClient client = KuttiDBClient.connectManaged(options)) {
            client.put("managed-java", "value".getBytes(StandardCharsets.UTF_8));
            byte[] value = client.get("managed-java");
            if (value == null || !"value".equals(new String(value, StandardCharsets.UTF_8)))
                throw new AssertionError("managed Java value mismatch");
        }
    }
}
